kernel/include/ksched.h

```diff
@@ -313,4 +323,134 @@ static inline struct k_thread *z_unpend1_no_timeout(_wait_q_t *wait_q)
 	return thread;
 }
 
+/*
+ * APIs for working with the Zephyr kernel scheduler. Intended for use in
+ * management of IPC objects, either in the core kernel or other IPC
+ * implemented by OS compatibility layers, providing basic wait/wake operations
+ * with spinlocks used for synchronization.
+ *
+ * These APIs are public and will be treated as contract, even if the
+ * underlying scheduler implementation changes.
+ */
+
+/**
+ * Callback function prototype for z_sched_wake_one/wake_all
+ *
+ * If a thread is woken up, the callback function is called with the thread
+ * as an argument. The sched_spinlock is held the entire time, ensuring that
+ * thread state doesn't suddenly change (from some nested IRQ or activity on
+ * another CPU) when the callback is modifying the thread's state.
+ *
+ * It is only inside these callbacks that it is safe to inspect or modify
+ * the thread that was woken up.
+ */
+typedef void (*z_sched_wake_cb_t)(struct k_thread *thread, void *obj,
+				  void *context);
+
+/**
+ * Wake up a thread pending on the provided wait queue, with callback
+ *
+ * Given a wait_q, wake up the highest priority thread on the queue. If the
+ * queue was empty just return false.
+ *
+ * Otherwise, do the following, in order,  holding sched_spinlock the entire
+ * time so that the thread state is guaranteed not to change:
+ * - if the callback function is provided, run the callback function with
+ *   the provided object and context pointers
+ * - Set the thread's swap return values to swap_retval and swap_data
+ * - un-pend and ready the thread, but do not invoke the scheduler.
```

Comment:
These details are all uniformly scheduler-specific.  What's the value of trying to document them in a header like this?  The only code that can productively use this is going to be effectively part of the scheduler internals anyway.  Note (mentioned elsewhere too) that holding the scheduler lock means that virtually the entirety of the Zephyr process state API is going to be unusable to this callback!

Wouldn't it be simpler to just call out to whatever functions you want and discriminate via an argument here?  The callback abstraction seems like a guaranteed footgun for someone in the future.